### DISCUSSION TITLE
Generators for tree sitter grammar

### DISCUSSION THREAD

**Discussion by will62794** [tlaplus-community/tree-sitter-tlaplus#57] [Q&A]
Can tree sitter grammars also be used to create generators i.e. output random strings that match the grammar? I know some similar projects like [nearleyjs](https://nearley.js.org/#features) have this feature. Being able to generate random TLA+ expressions could be quite helpful for certain testing scenarios e.g. [conformance checking](https://github.com/will62794/tla-web/blob/ee97b34713962e6a7303beaed90d4738dc23fe5a/specs/with_state_graphs/tla_expr_eval.tla) between the output of the [Javascript TLA+ interpreter](https://github.com/will62794/tla-web) and TLC.

**Comment by ahelwer** [comment]
No such functionality exists with tree-sitter, although a few people have talked about it (see https://github.com/tree-sitter/tree-sitter/issues/642 and https://github.com/tree-sitter/tree-sitter/discussions/1676). It would indeed be useful. Note that since the grammar is permissive (accepts some invalid TLA+ expressions) it might generate some invalid TLA+ expressions if not used carefully. I'm not sure how the external scanner would fit into any generator either.

**Comment by will62794** [comment]
@ahelwer Sounds good, thanks for the info. 

On a semi-related note, all of your work on the tree-sitter parser, and my recent work on writing a Javascript based TLA+ interpreter makes me wish there was a better unifying framework for this kind of problem. That is, an alternative to re-implementing this logic for parsing/interpreting in various contexts/languages. It would be nice if we could (formally) specify the behavior of TLA+ parsing/evaluation/execution once, and compile it down into various languages. I don't know if attempts at this kind of thing have been done over the years, but it seems a potentially interesting direction to think about, probably more so as a research/exploratory project. 

Defining systems [in an abstract language like Coq](https://homes.cs.washington.edu/~ztatlock/pubs/verdi-wilcox-pldi15.pdf) and extracting them to code is perhaps one approach to this kind of problem, but it might also bring a lot of extra complexity. My understanding was also that parsing TLA+, for example, is not as straightforward as defining say, a context-free grammar, so presumably you'd also need a somewhat general framework (e.g. not limited to what standard parser generators can do).

**Reply by ahelwer** [reply]
You're correct TLA+ is not a context-free language. Proofs and conjunction/disjunction lists are both context-sensitive, meaning the interpretation of the token sequence changes depending on previously-encountered tokens (column position of previous juncts in the case of junctlists, and enclosing proof levels in the case of proofs). These are implemented in the external scanner of my tree-sitter grammar, along with some tokens where I needed to override tree-sitter's lexing behavior.

There have been a lot of attempts over the years to create a standardized grammar DSL which can be consumed by multiple parser generators to produce their own parsers (like [YACC](https://en.wikipedia.org/wiki/Yacc)), but I don't think any of them were really successful. The state-of-the-art approach for writing really good parsers is to write them by hand, not even using a parser generator. This is because a lot of parser logic is devoted to providing good error messages (something tree-sitter does not do) which is very difficult to generalize. My plan at this time is to spend much of 2022 writing a TLA+ parser in Rust. I hope this will serve as a rock-solid basis for interpreters or model checkers and liberate us from the Java ecosystem.

I've never personally written an interpreter (although I did recently acquire [this book](http://craftinginterpreters.com/)) so I know considerably less about the interpreter landscape and whether that field has similarly gone through attempts at generalization. I know the [Racket](https://racket-lang.org/) programming language is often used in this area, though.

**Reply by will62794** [reply]
> The state-of-the-art approach for writing really good parsers is to write them by hand, not even using a parser generator. This is because a lot of parser logic is devoted to providing good error messages (something tree-sitter does not do) which is very difficult to generalize. My plan at this time is to spend much of 2022 writing a TLA+ parser in Rust. I hope this will serve as a rock-solid basis for interpreters or model checkers and liberate us from the Java ecosystem.

Yes, in general, we might just consider a parser to be an arbitrary program (i.e. a Turing machine). Of course, in practice, we at least want the language we're parsing to be decidable (e.g. see [1]). It seems TLA+ falls either into the _context-sensitive_ [2] or _recursive_ [3] bucket. I'm not sure if its exact classification could be formalized, but it's likely not something that would ever be worth the effort (e.g. I'm not really concerned with this formal classification, as long as we can write practically useful parsers for TLA+). 

Rather, I am interested in the prospect of a kind of "general algorithm language", in a sense, that can be used to describe general procedures (say, for parsing/interpreting TLA+) and then be compiled into different target languages. Parser generators, of course, have tried to solve this problem, but for a restricted subclass of problem (e.g. parsing context-free grammars). Overall, I'm definitely in favor of having different parser/interpreter implementations within a language ecosystem, but having to, for example, re-implement and re-test all the TLA+ parsing features in Rust (or the interpreter facilities in Javascript) always feels like a duplication of work at some fundamental level. 

In essence, I feel like the "core parsing algorithm" for TLA+ should be able to be formalized in a way that is abstract enough to be portable across various languages, but concrete enough that it can be translated to real code in different programming languages. Perhaps even in a way that generates very efficient implementations in a target language. Maybe we could go further, cutting out the target language and just compile directly to something like LLVM or WebAssembly. Not sure.

Note that this is more of a theoretical question/problem, but it's just something that came up in my mind when I was thinking about the various re-implementations of core TLA+ tools in different contexts.

[1] https://blog.reverberate.org/2013/08/parsing-c-is-literally-undecidable.html
[2] https://en.wikipedia.org/wiki/Context-sensitive_language
[3] https://en.wikipedia.org/wiki/Recursive_language

**Reply by ahelwer** [reply]
I think we can distinguish two things here:

1. Specifying the core parsing algorithm so people can understand/adapt it themselves instead of having to infer/reinvent it like I did
2. Machine-translating the grammar specification into usable code

TLA+ works *okay* as a specification language for CFGs (see Lamport's TLA+ grammar in TLA+ [here](http://lamport.azurewebsites.net/tla/TLAPlus2Grammar.tla)) but I don't know how you would do the same for CSGs. Lamport chose to specify the context-free part of TLA+ in TLA+ while documenting the context-sensitive parts (and the precedence parts) in plain language (in *Specifying Systems* with V2 specified [here](http://lamport.azurewebsites.net/tla/tla2.html)).

It's also possible that Racket already implements all of this. I haven't looked into it too much.

**Reply by ahelwer** [reply]
Somewhat related, I recently found out that Stephan Merz formalized a bunch of TLA+ in Isabelle: https://github.com/tlaplus/tlapm/tree/main/isabelle
